x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/Case.java

```diff
@@ -439,7 +462,73 @@ public void close() {
 
         @Override
         public String toString() {
-            return "CaseEvaluator[conditions=" + conditions + ", elseVal=" + elseVal + ']';
+            return "CaseLazyEvaluator[conditions=" + conditions + ", elseVal=" + elseVal + ']';
+        }
+    }
+
+    private record CaseEagerEvaluatorFactory(
+        ElementType resultType,
+        ConditionEvaluatorSupplier conditionFactory,
+        ExpressionEvaluator.Factory elseValueFactory
+    ) implements ExpressionEvaluator.Factory {
+        @Override
+        public ExpressionEvaluator get(DriverContext context) {
+            ConditionEvaluator conditionEvaluator = conditionFactory.apply(context);
+            ExpressionEvaluator elseValue = null;
+            try {
+                elseValue = elseValueFactory.get(context);
+                ExpressionEvaluator result = new CaseEagerEvaluator(resultType, context.blockFactory(), conditionEvaluator, elseValue);
+                conditionEvaluator = null;
+                elseValue = null;
+                return result;
+            } finally {
+                Releasables.close(conditionEvaluator, elseValue);
+            }
+        }
+
+        @Override
+        public String toString() {
+            return "CaseEagerEvaluator[conditions=[" + conditionFactory + "], elseVal=" + elseValueFactory + ']';
+        }
+    }
+
+    private record CaseEagerEvaluator(
+        ElementType resultType,
+        BlockFactory blockFactory,
+        ConditionEvaluator condition,
+        EvalOperator.ExpressionEvaluator elseVal
+    ) implements EvalOperator.ExpressionEvaluator {
+        @Override
+        public Block eval(Page page) {
+            try (
+                BooleanBlock lhsOrRhsBlock = (BooleanBlock) condition.condition.eval(page);
+                ToMask lhsOrRhs = lhsOrRhsBlock.toMask();
+                Block lhs = condition.value.eval(page);
+                Block rhs = elseVal.eval(page);
+                Block.Builder builder = resultType.newBlockBuilder(lhs.getTotalValueCount(), blockFactory)
+            ) {
+                if (lhsOrRhs.hadMultivaluedFields()) {
+                    condition.registerMultivalue();
+                }
+                for (int p = 0; p < lhs.getPositionCount(); p++) {
```

Comment:
We could leverage `BooleanVector.isConstant()` here, eg.

```
                BooleanVector mask = lhsOrRhs.mask();
                if (mask.isConstant()) {
                    builder.copyFrom(mask.getBoolean(0) ? lhs : rhs, 0, lhs.getPositionCount());
                } else {
                    for (int p = 0; p < lhs.getPositionCount(); p++) {
                    ...
```
Or even return lhs/rhs...? Not sure it's safe for memory accounting

Reply:
I believe it is safe, but tests can figure it out. I'll try it.